### ISSUE TITLE
Define routes for nodes in one place

### ISSUE THREAD

**Issue by hjvedvik** [gridsome/gridsome#455]
Move route options for all content types into a `routes` _(or any better name)_ section in `gridsome.config.js`. Currently, each source plugin implements routes on their own. But after this change, source plugins doesn't need to worry about routing etc. We also believe it will give a better overview having all node routes in one place. And it will be easier to extend the functionality later. Content types not defined there will not be generated or have a `path` field in the GraphQL schema.

The `routes` config will be config for how pages for nodes should be generated. This is not a breaking change, but the `route` option for content types will be deprecated.

So instead of defining a `route` for the source or a content type like this:

```js
{
  use: '@gridsome/source-filesystem',
  options: {
    typeName: 'User',
    route: '/user/:name'
  }
}
```

Define the route in `routes` like this:

```js
module.exports = {
  routes: {
    User: '/user/:name' // typeName: route
  }
}
```

Here is how the full config might look like:

```js
module.exports = {
  routes: {
    // simple config
    Post: '/:year/:month/:day/:slug',
    // specify component
    Post: {
      path: '/:year/:month/:day/:slug',
      component: './src/templates/CustomPostTemplate.vue'
    },
    // all possible options
    Post, {
      path: '/:year/:month/:day/:slug',
      component: './src/templates/Post.vue',
      dynamicRoute: false, // generate one route for each node
      fieldName: 'path' // graphql field name
    },
    // generate multiple templates for content type
    Author: [
      {
        path: '/author/:name',
        component: './src/templates/Author.vue'
      },
      {
        path: '/author/:name/books',
        component: './src/templates/AuthorBooks.vue',
        fieldName: 'booksPath'
      }
    ]
  }
}
```

**Comment by Al-Rozhkov** [comment]
Would it be possible to set up nested routes, dynamic nested routes or named views?

**Comment by tomtev** [comment]
@Al-Rozhkov 0.7 will have client-side route support. We're thinking about doing it as Nuxt does. Where you add for example a `user/_id.vue` file to get a `user/:id` route + an option to do it programmatically.

**Comment by Atinux** [comment]
I like this idea of having all the routes in the same key, more readability 👌

About routing, we plan to move from pages/_id.vue to pages/[id].vue for handling multiple params and better readability too :)

**Comment by hjvedvik** [comment]
I think `pages/[id].vue` looks much better @Atinux :) We'll consider the same for the same reasons.

@Al-Rozhkov Nested routes and named views could be interesting. I'm not sure how to implement it with automatic generation and GraphQL etc atm, but will definitely take a look at it sometime.

**Comment by igregson** [comment]
> @Al-Rozhkov 0.7 will have client-side route support. We're thinking about doing it like Nuxt does. Where you add for example a `user/_id.vue` file to get a `user/:id` route + an option to do it programmatically.

Did this land in the recently-released 0.7? I see that "dynamic routing" did #570), but the part there about "[...] are generated as static HTML files like normal pages [...]" makes it sound like this doesn't include client-side only routing (where required data for a route is only known at runtime, on the client).

**Comment by hjvedvik** [comment]
@igregson Yes, its included in 0.7. The static HTML part means that the component for the dynamic route will generate one static HTML file. It is only meant to be used in rewrite rules. For example, `/user/:id` will have a file at `/user/_id.html` which must have rewrite rule on the server to work if someone visits, for example, `/user/123` directly.

**Comment by soongsta** [comment]
Did the routes feature get shipped? If so, is there any instructions on how to try this out?
I really like the idea of being able to do the following:

<pre><code>
Author: [
      {
        path: '/author/:name',
        component: './src/templates/Author.vue'
      },
      {
        path: '/author/:name/books',
        component: './src/templates/AuthorBooks.vue',
        fieldName: 'booksPath'
      }
    ]
</code></pre>
